stdlib/public/core/StringNormalization.swift

```diff
@@ -102,3 +112,331 @@ internal func _tryNormalize(
   }
   return numericCast(count)
 }
+
+internal struct NormalizationResult {
+  var amountFilled: Int
+  var nextReadPosition: String.Index
+  var reallocatedBuffers: Bool
+}
+
+private func fastFill(
+  _ sourceBuffer: UnsafeBufferPointer<UInt8>,
+  _ outputBuffer: UnsafeMutableBufferPointer<UInt8>
+) -> (Int, Int) {
+  // Quick check if a scalar is NFC and a segment starter
+  @inline(__always) func isNFCStarter(_ scalar: Unicode.Scalar) -> Bool {
+    // Fast-path: All scalars up through U+02FF are NFC and have boundaries
+    // before them
+    if scalar.value < 0x300 { return true }
+
+    // Otherwise, consult the properties
+    return scalar._hasNormalizationBoundaryBefore && scalar._isNFCQCYes
+  }
+  
+  let outputBufferThreshold = outputBuffer.count - 4
+  
+  // TODO: Additional fast-path: All CCC-ascending NFC_QC segments are NFC
+  // TODO: Just freakin do normalization and don't bother with ICU
+  var outputCount = 0
+  let outputEnd = outputBufferThreshold
+  var inputCount = 0
+  let inputEnd = sourceBuffer.count
+  while inputCount < inputEnd && outputCount < outputEnd {
+    // TODO: Slightly faster code-unit scan for latiny (<0xCC)
+
+    // Check scalar-based fast-paths
+    let (scalar, len) = _decodeScalar(sourceBuffer, startingAt: inputCount)
+    _internalInvariant(inputCount &+ len <= inputEnd)
+
+    if _slowPath(
+         !sourceBuffer.hasNormalizationBoundary(before: inputCount &+ len)
```

Comment:
Can probably drop this check, but please verify

Reply:
We can't drop this. `_isNFCStarter` checks if the current scalar has a boundary before. This is checking if there's a boundary before the scalar after this.